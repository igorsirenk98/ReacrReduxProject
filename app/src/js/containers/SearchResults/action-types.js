export const FETCH_PRODUCTS_BY_SEARCH_START = 'FETCH_PRODUCTS_BY_SEARCH_START';
export const FETCH_PRODUCTS_BY_SEARCH_RESULT = 'FETCH_PRODUCTS_BY_SEARCH_RESULT';