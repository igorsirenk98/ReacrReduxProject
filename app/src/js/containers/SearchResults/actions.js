import {
	FETCH_PRODUCTS_BY_SEARCH_START,
	FETCH_PRODUCTS_BY_SEARCH_RESULT
} from './action-types';

import callApi from '../../utils/callApi';

export const fetchProductsBySearchStart = () => ({
	type: FETCH_PRODUCTS_BY_SEARCH_START,
});

export const fetchProductsBySearchResult = (payload, error) => ({
	type: FETCH_PRODUCTS_BY_SEARCH_RESULT,
	payload,
	error
});

export function fetchProductsBySearch(searchValue) {
	const url = `products/search=${searchValue}`;

	return (dispatch) => {
		return dispatch(callApi([fetchProductsBySearchStart, fetchProductsBySearchResult], url));
	}
}