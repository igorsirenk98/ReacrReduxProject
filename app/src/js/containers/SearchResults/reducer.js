import {
	FETCH_PRODUCTS_BY_SEARCH_START,
	FETCH_PRODUCTS_BY_SEARCH_RESULT
} from './action-types';

const initialState = {
	payload: {
		products: []
	},
	error: false
};

const searchResults = (state = initialState, action) => {
	switch (action.type) {
		case FETCH_PRODUCTS_BY_SEARCH_START: {
			return {
				...state,
				error: false,
			};
		}
		case FETCH_PRODUCTS_BY_SEARCH_RESULT: {
			return {
				...state,
				payload: action.payload,
				error: action.error
			};
		}
		default:
			return state;
	}
};

export default searchResults;
