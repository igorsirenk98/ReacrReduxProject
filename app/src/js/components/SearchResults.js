import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

import Button from '@material-ui/core/Button';
import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';

import { List } from './basicComponents/List';
import Loading from './basicComponents/Loading';
import { fetchProductsBySearch } from '../actions/ProductsBySearch';

const mapStateToProps = state => ({
    searchedProducts: state.productsBySearch.searchedProducts,
    loading: state.productsBySearch.loading,
    error: state.productsBySearch.error
});

class SearchResults extends Component {
    componentDidUpdate(prevProps) {
        const { searchInputValue: currentSearchInputValue } = this.props.match.params;
        const { searchInputValue: prevSearchInputValue } = prevProps.match.params;

        if (currentSearchInputValue !== prevSearchInputValue) {
            this.props.dispatch(fetchProductsBySearch(currentSearchInputValue));
        }
    }

    componentDidMount() {
        const { searchInputValue } = this.props.match.params;

        this.props.dispatch(fetchProductsBySearch(searchInputValue));
    }

    render() {
        const { searchedProducts, loading, error } = this.props;

        if (error) {
            return <div className="error"><p>Error! {error.message}</p></div>;
        } else if (loading) {
            return (
                <>
                    <CssBaseline />
                    <Container className="containerStyles" fixed>
                        <div className="loadingContainer">
                            <Loading />
                        </div>
                    </Container>
                </>
            );
        }

        return (
            <>
                {!!searchedProducts.length ? (
                    <List products={searchedProducts}/>
                ) : (
                    <Typography class="error" gutterBottom variant="h5" component="h2">
                        There is no results for "{this.props.match.params.searchInputValue}"
                    </Typography>
                )
                    
                }
            </>
        ); 
    }
}

export default connect(mapStateToProps)(SearchResults);