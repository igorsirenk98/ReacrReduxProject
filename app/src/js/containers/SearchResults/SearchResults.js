import React, { Component } from 'react';

import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import CssBaseline from '@material-ui/core/CssBaseline';
import Typography from '@material-ui/core/Typography';
import Container from '@material-ui/core/Container';

import List from '../../components/controls/List';
import ErrorMessage from '../../components/controls/ErrorMessage';
import Loading from '../controls/Loading/Loading';

import { fetchProductsBySearch } from './actions';

const mapStateToProps = state => ({
    payload: state.productsBySearch.payload,
    isLoading: state.loadingAnimation.payload.isLoading,
    error: state.productsBySearch.error,
});

const mapDispatchToProps = dispatch => {
    return bindActionCreators({
        fetchProductsBySearch
    }, dispatch);
};

export class SearchResults extends Component {
    componentDidUpdate(prevProps) {
        const { searchInputValue: currentSearchInputValue } = this.props.match.params;
        const { searchInputValue: prevSearchInputValue } = prevProps.match.params;

        if (currentSearchInputValue !== prevSearchInputValue) {
            this.props.fetchProductsBySearch(currentSearchInputValue);
        }
    }

    componentDidMount() {
        const { searchInputValue } = this.props.match.params;

        this.props.fetchProductsBySearch(searchInputValue);
    }

    render() {
        const {
            payload,
            isLoading,
            error
        } = this.props;
        const { products } = payload;
        const { searchInputValue } = this.props.match.params;

        if (error) {
            return <ErrorMessage error={payload} />;
        } else if (isLoading) {
            return (
                <>
                    <CssBaseline />
                    <Container className="container" fixed>
                        <div className="loading-container">
                            <Loading />
                        </div>
                    </Container>
                </>
            );
        }

        if (!!products.length) {
            return (
                <>
                    <Typography className="notification" gutterBottom variant="h3" component="h2">
                        Results for "{decodeURIComponent(searchInputValue)}"
                    </Typography>
                    <List products={products} />
                </>
            );
        } else {
            return (
                <>
                    <CssBaseline />
                    <Container className="container" fixed>
                        <Typography className="notification" gutterBottom variant="h3" component="h2">
                            There is no results for "{decodeURIComponent(searchInputValue)}"
                    </Typography>
                    </Container>
                </>
            );
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(SearchResults);