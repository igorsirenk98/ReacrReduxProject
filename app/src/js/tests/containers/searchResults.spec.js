import React from 'react';
import { shallow } from 'enzyme';
import { SearchResults } from '../../containers/SearchResults/SearchResults';

describe('SearchResults container', () => {
    const props = {
        payload: {
            products: [{name: 'bike1'}, {name: 'bike2'}]
        },
        match: {
            params: 1
        },
        isLoading: false,
        error: false,
        fetchProductsBySearch: () => { },
    }

    describe('SearchResults initial', () => {
        const searchResults = shallow(<SearchResults {...props} />);

        it('renders correctly', () => {
            expect(searchResults.debug()).toMatchSnapshot();
        })
    })

    describe('SearchResults isLoading', () => {
        const nextProps = {
            ...props,
            payload: {
                products: [{name: 'bike1'}, {name: 'bike2'}],
            },
            isLoading: true,
        }

        const searchResults = shallow(<SearchResults {...nextProps} />);

        it('renders correctly', () => {
            expect(searchResults.debug()).toMatchSnapshot();
        })
    })

    describe('SearchResults error', () => {
        const nextProps = {
            ...props,
            payload: {
                error: {
                    message: 'error',
                }
            },
            error: true,
        }

        const searchResults = shallow(<SearchResults {...nextProps} />);

        it('renders correctly', () => {
            expect(searchResults.debug()).toMatchSnapshot();
        })
    })

    describe('SearchResults with data', () => {
        const nextProps = {
            ...props,
            payload: {
                products: [{name: 'bike1'}, {name: 'bike2'}],
            }
        }

        const searchResults = shallow(<SearchResults {...nextProps} />);

        it('renders correctly', () => {
            expect(searchResults.debug()).toMatchSnapshot();
        })
    })
})

