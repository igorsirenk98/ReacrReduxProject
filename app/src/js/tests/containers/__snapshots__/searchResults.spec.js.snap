// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchResults container SearchResults error renders correctly 1`] = `"<ErrorMessage error={{...}} />"`;

exports[`SearchResults container SearchResults initial renders correctly 1`] = `
"<Fragment>
  <WithStyles(ForwardRef(Typography)) className=\\"notification\\" gutterBottom={true} variant=\\"h3\\" component=\\"h2\\">
    Results for &quot;
    undefined
    &quot;
  </WithStyles(ForwardRef(Typography))>
  <List products={{...}} />
</Fragment>"
`;

exports[`SearchResults container SearchResults isLoading renders correctly 1`] = `
"<Fragment>
  <CssBaseline />
  <WithStyles(ForwardRef(Container)) className=\\"container\\" fixed={true}>
    <div className=\\"loading-container\\">
      <Connect(Loading) />
    </div>
  </WithStyles(ForwardRef(Container))>
</Fragment>"
`;

exports[`SearchResults container SearchResults with data renders correctly 1`] = `
"<Fragment>
  <WithStyles(ForwardRef(Typography)) className=\\"notification\\" gutterBottom={true} variant=\\"h3\\" component=\\"h2\\">
    Results for &quot;
    undefined
    &quot;
  </WithStyles(ForwardRef(Typography))>
  <List products={{...}} />
</Fragment>"
`;
